src/NetEscapades.EnumGenerators/SourceGenerationHelper.cs

```diff
@@ -962,4 +963,63 @@ public static string[] GetNames()
             .ToString();
         return (content, filename);
     }
+
+    private static IReadOnlyCollection<(string Key, EnumValueOption Value)> GetNamesOrderedByValue(EnumToGenerate enumToGenerate)
+    {
+        var valueComparer = Comparer<object>.Default;
+        var tupleComparer = Comparer<(string Key, EnumValueOption Value)>
+            .Create((x, y) => valueComparer.Compare(x.Value.ConstantValue, y.Value.ConstantValue));
+
+        var names = enumToGenerate.Names;
+        var i = 1;
+        for (; i < names.Count; i++)
+        {
+            if (tupleComparer.Compare(names[i-1], names[i]) > 0)
+            {
+                // ConstantValue i is bigger than ConstantValue[i-1]
+                break;
+            }
+        }
+
+        if (i == names.Count)
+        {
+            // Everything is already ordered
+            return names;
+        }
+
+        var orderedNames = new (string Key, EnumValueOption Value)[names.Count];
+        // Copy the already ordered values
+        names.AsSpan()[..i].CopyTo(orderedNames);
+        
+        for (; i < names.Count; i++)
+        {
+            var targetPos = Array.BinarySearch(orderedNames, 0, i, names[i], tupleComparer);
+            if (targetPos < 0)
+            {
+                // No match, after bit-inverting we are on the lowest bigger entry
+                targetPos = ~ targetPos;
+            }
+            else
+            {
+                // Match, so we are on an equal item. Move to the lowest bigger entry
+                while (targetPos < i)
+                {
+                    targetPos++;
+                    if (tupleComparer.Compare(orderedNames[targetPos], names[i]) > 0)
+                    {
+                        break;
+                    }
+                }
+            }
+
+            if (targetPos < i)
+            {
+                Array.Copy(orderedNames, targetPos, orderedNames, targetPos + 1, i - targetPos);
```

Comment:
This all seems quite expensive 🤔 Can't we just use a "normal" Linq `OrderBy` 🤔 It doesn't seem worth the extra complexity here? And we don't need to be outputting an `EquatableArray`, a normal `List` or array would do at this stage I think

Reply:
It should be simpler than Linq's `OrderBy`, although I agree, that it is more explicit. The effect is the same as e.g.
```cs
enumToGenerate.Names
            .Select((n, i) => (n, i))
            .OrderBy(t => t.n.Value.ConstantValue)
            .ThenBy(t => t.i)
            .Select(t => t.n)
            .ToList();
```
which is ordering the names by underlying values, keep names for the same underlying value in the order that they were passed to us. I used the `EquatableArray`, as that was the type previously used and I thought it served some purpose.

Reply:
Cool, thanks, I think that code is _much_ clearer! 😄 (though as a nit, I would prefer `name` over `n` 😉)

> I used the EquatableArray, as that was the type previously used and I thought it served some purpose.

Totally, it serves an important part in the _core_ of the generator, where instances have to satisfy equality to ensure we get the correct incremental caching. 

But once we've got to this point, it's not necessary - the instances are going to be thrown away straight after being built, so just being the most efficient here is as good as anything. 

_Ordinarily_ I would say _don't_ use Linq because of that, but IMO the improvement in readability is definitely worth it here. IMO there's no need for unit tests with the Linq statement - it's clear what it does from a glance 🙂

Reply:
I thought it would read ok with the comments, but I don't really care that much. Guess it's a bunch more allocations, but the result is most important to me